README.md

```diff
@@ -1 +1,329 @@
+            0x59, 0xaf, 0x33, 0xdc, 0xc5,
+        ]),
+    ),
+]);
+```
+
+Note that the IPLD parser has dropped the `role: "admin"` key.
+
+The `"sig"` field is then removed, and the remaining fields serialized to binary;
+
+``` Rust
+Ipld::DagJson::serialize(
+    Ipld::Assoc([
+        ("role", Ipld::String("user")),
+        (
+            "links",
+            Ipld::Array([
+                Ipld::Cid("bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"),
+                Ipld::Cid("bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"),
+                Ipld::Cid("bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"),
+            ]),
+        )
+    ])
+);
+```
+
+The signature is then checked against the above fields, which passes since there's only a `role: "user"` entry. The application then uses the original JSON with the `role: "admin"` entry.
+
+# 2 Safety
+
+Data that has already been parsed to an in-memory IPLD representation can be canonically encoded trivially: it has already been through a [parser / validator][Parse Don't Validate].
+
+Data purporting to conform to an IPLD encoding (such as [DAG-JSON]) MUST be validated prior to signature verification. This MAY be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signalled if it does not match.
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec][DAG-JSON]
+
+As it is critical for guarding against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+### 2.1.1 Caching & Invalidation
+
+Signing CIDs has two additional caching effects
+
+* Signing CIDs enables a simple strategy for caching validation by CID
+* Such a strategy also requires accounting for revocation of the signing keys, and so need to be tagged with this additional metadata
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x55`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+After being decoded from [unsigned varint]s, a varsig includes the following segments:
+
+```abnf
+varsig = multibase-prefix %x34 varsig-header varsig-body
+multibase-prefix = ALPHA ; Multibase
+varsig-header = unsigned-varint ; Usually the public key code from Multicodec
+varsig-body = *unsigned-varint; Zero or more segments required by the kind of varsig (e.g. raw bytes, hash algorithm, etc)
```

Comment:
> For security, to avoid memory attacks, we use a "practical max" of 9 bytes.

Ah, I didn't realize this! Sure, let's add a length parameter 👍

Reply:
...or it could be set by the varsig-header. This is what `did:key` does, for example:

<img width="797" alt="Screenshot 2023-03-24 at 12 58 43" src="https://user-images.githubusercontent.com/1052016/227626998-59aacc22-db89-4c97-b293-5dcf1da8dc56.png">

So RSA would need to set the body lenght, but EdDSA is fixed and doesn't need to include it

Reply:
I think that's fine as well!